types/src/transaction/mod.rs

```diff
@@ -1115,31 +1105,97 @@ impl TransactionListWithProof {
                 event_lists.len(),
                 self.transactions.len(),
             );
-            itertools::zip_eq(event_lists, self.proof.transaction_infos())
-                .map(|(events, txn_info)| {
-                    let event_hashes: Vec<_> = events.iter().map(ContractEvent::hash).collect();
-                    let event_root_hash =
-                        InMemoryAccumulator::<EventAccumulatorHasher>::from_leaves(&event_hashes)
-                            .root_hash();
-                    ensure!(
-                        event_root_hash == txn_info.event_root_hash(),
-                        "Some event root hash calculated doesn't match that carried on the \
-                         transaction info.",
-                    );
-                    Ok(())
-                })
+            itertools::zip_eq(event_lists, &self.proof.transaction_infos)
+                .map(|(events, txn_info)| verify_events_against_root_hash(events, txn_info))
                 .collect::<Result<Vec<_>>>()?;
         }
 
         Ok(())
     }
+}
 
-    fn display_option_version(version: Option<Version>) -> String {
-        match version {
-            Some(v) => format!("{}", v),
-            None => String::from("absent"),
+/// This differs from TransactionListWithProof in that TransactionOutputs are
+/// stored (no transactions). Events are stored inside each TransactionOutput.
+///
+/// Note: the proof cannot verify the TransactionOutputs themselves. This
+/// requires speculative execution of each TransactionOutput to verify that the
+/// resulting state matches the expected state in the proof (for each version).
+#[derive(Clone, Debug, Eq, PartialEq, Deserialize, Serialize)]
+pub struct TransactionOutputListWithProof {
+    pub transaction_outputs: Vec<TransactionOutput>,
+    pub first_transaction_output_version: Option<Version>,
+    pub proof: TransactionInfoListProof,
+}
+
+impl TransactionOutputListWithProof {
+    pub fn new(
+        transaction_outputs: Vec<TransactionOutput>,
+        first_transaction_output_version: Option<Version>,
+        proof: TransactionInfoListProof,
+    ) -> Self {
+        Self {
+            transaction_outputs,
+            first_transaction_output_version,
+            proof,
         }
     }
+
+    /// A convenience function to create an empty proof. Mostly used for tests.
+    pub fn new_empty() -> Self {
+        Self::new(vec![], None, TransactionInfoListProof::new_empty())
+    }
+
+    /// Verifies the transaction output list with proof using the given `ledger_info`.
+    /// This method will ensure:
+    /// 1. All transaction infos exist on the given `ledger_info`.
+    /// 2. If `first_transaction_output_version` is None, the transaction output list is empty.
+    ///    Otherwise, the list starts at `first_transaction_output_version`.
+    /// 3. Events in each transaction output match the expected event root hashes in the proof.
+    pub fn verify(
```

Comment:
nit: repeat the note/caveat here about not verifying the TransactionOutputs, just the TransactionInfos and events